File: 200_LeetCode_Questions/D10P2_LeetCode_34.cpp
// The code modies the working of binary search by not stopping at where the element is found rather it pushes to left or right depending upon the result
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int l,r,m,s,e;
        l=0;
        r=nums.size();
        if(r<=l)
            return vector<int>{-1,-1};
        r=r-1;
        s=-1;
        while(l<=r){
            m=(l+r)/2;
        if(nums[m]>target){
                r=m-1;
            }else if(nums[m]<target){
                l=m+1;
            }else{
                r=m-1;
                s=m;
            }
         }
        l=0;
        e=-1;
         r=nums.size()-1;
         while(l<=r){
            m=(l+r)/2;
            if(nums[m]>target){
                r=m-1;
            }else if(nums[m]<target){
                l=m+1;
            }else{
                l=m+1;
                e=m;
            }
        }
        return vector<int>{s,e};
    }
};

//Time Complexity = O(log(n)+log(n))= O(log(n))

File: 200_LeetCode_Questions/Python/D10P2_LeetCode_34.py
# Find First and Last Position of Element in Sorted Array
class Solution(object):
    def searchRange(self, nums, target):
        if len(nums) == 0: 
            return([-1, -1])
        low_element, high_element = 0, len(nums)
        while low_element < high_element:
            mid_element = (high_element + low_element) // 2
            if nums[mid_element] == target: 
                left_element, right_element = mid_element, mid_element
                while (left_element-1) >= 0 and nums[left_element-1] == target: 
                    left_element -= 1
                while (right_element+1) < len(nums) and nums[right_element+1] == target: 
                    right_element += 1
                return([left_element, right_element])
                break
            elif nums[mid_element] < target: 
                low_element = mid_element + 1
            elif nums[mid_element] > target: 
                high_element = mid_element
        return([-1, -1])